### PR TITLE
Add SteHeaviside quantizer

### PR DESCRIPTION
**Pull request by MariaHeuss**
This adds the quantizer for the AND quantization scheme. The name ReluSign is open for discussion. The default parameters might not be the best but since there were not enough experiments on this yet and this works I would keep it like this for now.

### CODE REVIEW

File: larq/quantizers.py [hunk 1]
```diff
@@ -450,6 +452,90 @@ def __init__(
...
+    1 & \left|x\right| \leq 1 \\\
+    0 & \left|x\right| > 1
+    \end{cases}\\]
+
+    ```plot-activation
+    quantizers.relu_sign
+    ```
+
+    # Arguments
+    threshold_value: The value for the threshold, $\Delta$, default value 0.
```
[the remark this answers is not in the document]
Comment (MariaHeuss) [reply larq/quantizers.py]: Probably true, I added it to be able to experiment with it but we I would expect the batch norm to do the work as well. But don't you thing there might be situations where there is no batch norm or bias directly before the quantizer? I am also happy to delete the threshold value until we find that we do need it after all.

File: larq/quantizers.py [hunk 2]
```diff
@@ -450,6 +452,90 @@ def __init__(
...
+    AND Binarization function
+    """
+
+    def __init__(self, threshold_value=0.0, clip_value=1.0):
+        super().__init__(relu_sign, threshold_value=threshold_value, clip_value=clip_value)
+
+
+@utils.register_keras_custom_object
+@utils.set_precision(1)
+def relu_sign(x, threshold_value=0.0, clip_value=1.0):
```
[the remark this answers is not in the document]
Comment (MariaHeuss) [reply larq/quantizers.py]: I am happy with going with ste_heaviside. ste_step and ste_unit_step I do not like since I would associate the sign function with it (especially unit_step seems just a bad name from the mathematical point of view since 0 is not a unit, I am surprised that people gave it this name).

File: larq/quantizers.py [hunk 3]
```diff
@@ -450,6 +452,90 @@ def __init__(
...
+    clip_value: Threshold for clipping gradients.
+
+    # Returns
+    AND-binarized tensor.
+    """
+    x = tf.clip_by_value(x, -clip_value, clip_value)
+
+    @tf.custom_gradient
+    def _and_binarize_with_identity_grad(x):
+        def grad(dy):
```
[the remark this answers is not in the document]
Comment (MariaHeuss) [reply larq/quantizers.py]: This is not something I found in the literature, it is something that I put together myself. Do we need a reference?

File: larq/quantizers.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (lgeiger) [reply larq/quantizers.py]: I think having a non-trainable offset is quite brittle and I guess would require a lot of hand tuning to get right. A bias can be learned easily and usually is added before the quantizer by default if not batch norm is used.
I think having this threshold could lead to some quite hard to debug training issues. E.g.:
Let's say we have a batch norm layer  right before the quantizer (let's neglect the bias in the batch norm for now). ~68% of the input to the quantizer will be between [-1, +1]. Let's say we keep the clipping at {-1, +1}. If we now manually set threshold to 2 or larger we will have no gradients for most of the values coming into the quantizer, which will lead to almost no learning here. Or am I missing something?

File: larq/quantizers.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lgeiger) [reply larq/quantizers.py]: Sure, I'm fine with `ste_heaviside`

File: larq/quantizers.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (lgeiger) [reply larq/quantizers.py]: No, we don't need a reference. I was just curious, since I'd expect the pseudo-gradient to be more an approximation of the sigmoid gradient rather then an approximation of the hard_tanh gradient.

File: larq/quantizers.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (MariaHeuss) [reply larq/quantizers.py]: You are right that you can just have a bias before the quantizer. The example is correct as well, and I agree that tuning the threshold is hard and might lead to some hard to debug training issues. I wanted to give the user some flexibility to try out stuff, but I agree that we can take away this option and add it again if we ever think it is needed after all.

File: larq/quantizers.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (MariaHeuss) [reply larq/quantizers.py]: I did not experiment with having a different gradient approximation. There are different possibilities for approximating the gradient. I picked this one because it is the easiest and the clipped equivalent of the ste of the hard_tanh would be something like a scaled version of this with a wider clip interval. So it does not seem like there would be a big difference. Anyways I think we want to spend some time on looking closer into AND quantization and its gradient. We can put this PR on hold in case you think this needs to be done first.

**Comment by MariaHeuss** [comment]
@lgeiger do you know why the larq.larq check is failing?

**Comment by lgeiger** [comment]
> please run black code formatter, to make CI happy.

**Comment by MariaHeuss** [comment]
Do you think we can merge this or would you rather wait until there are more experiments for the right choice of pseudo gradient?

File: larq/quantizers.py [hunk 4]
```diff
@@ -450,6 +452,86 @@ def __init__(
...
+            return dy
+
+        return (tf.sign(tf.sign(x) - 0.1) + 1) / 2, grad
```
Comment (lgeiger) [inline larq/quantizers.py]: Why not:
```suggestion
        return tf.sign(tf.nn.relu(x)), grad
```
I think this is a lot clearer and also might be faster?

**Comment by lgeiger** [comment]
> Do you think we can merge this or would you rather wait until there are more experiments for the right choice of pseudo gradient?

I think it's fine for now. Since it is called `ste_heaviside` we can always add an other quantizer that has another gradient.

File: larq/quantizers.py (hunk 4 above)
Comment (MariaHeuss) [reply larq/quantizers.py]: I was just imitating the ste_sign binarization, I thought you tested this one on speed.

File: larq/quantizers.py [hunk 5]
```diff
@@ -450,6 +452,86 @@ def __init__(
...
+    AND-binarized tensor.
+    """
+    x = tf.clip_by_value(x, -clip_value, clip_value)
+
+    @tf.custom_gradient
+    def _and_binarize_with_identity_grad(x):
+        def grad(dy):
+            return dy
+
+        return tf.nn.relu(tf.sign(x)), grad
```
Comment (lgeiger) [inline larq/quantizers.py]: I'd prefer the following, since TensorFlow might be able to do some fusing of the convolution and the ReLU. I am not sure though, but just in case:
```suggestion
        return tf.sign(tf.nn.relu(x)), grad
```
